### DISCUSSION TITLE
Parameter's modifier "this" for non static Method

### DISCUSSION THREAD

**Discussion by VanKrock** [dotnet/csharplang#1650] [Language Ideas] [answered]
Why "this" modifier used in extensions only? What about  use "this" for all types?

```csharp
    public class Obj
    {
        public bool Equals<T>(this T @this, T obj)
        {
            return @this == obj;
        }
    }
```

Using
```csharp
            var obj = new Obj();
            var secondObj = new Obj();
            obj.Equals(secondObj);
```

Its' can be translated to
```csharp
    public class Obj
    {
        public bool Equals<T>(T @this,T obj) where T : Obj
        {
            return @this == obj;
        }
    }
```

Using can be translated to
```csharp
            var obj = new Obj();
            var secondObj = new Obj();
            obj.Equals(obj, secondObj);
```

it's can be used for real fluent interfaces
```csharp
    public class ObjWithValue<T> : Obj, IHasValue<T>, ICanSet<T>
    {
        private T value;
        public T Value => value;

        public TObj Set<TObj>(this TObj @this, T value)
        {
            this.value = value;
            return @this;
        }
    }

    public interface IHasValue<T>
    {
        T Value { get; }
    }

    public interface ICanSet<T>
    {
        TObj Set<TObj>(this TObj @this, T value);
    }
```

**Comment by Joe4evr** [comment]
Instance methods already operate on the `this` instance, so what benefit would this bring?

**Comment by tannergooding** [comment, marked as answer]
Why not just use the `this` keyword directly?

For extension methods, it is used carry the object context and identifies the type the extension exists for.

For instance methods, the object context is implicitly the first parameter (and is abstracted away to the `this` keyword. The type that the method belongs to is identified by the enclosing code.

**Comment by VanKrock** [comment]
You can not return instance of Deprecated class from Base class.
You can not do this in C#
```csharp
        public TObj Set<TObj>(this TObj @this, T value)
        {
            this.value = value;
            return @this;
        }
```

**Comment by HaloFour** [comment]
@VanKrock 

The signature of instance methods is fixed by the CLR.  The implicit "this" argument is not specified in IL, and C# cannot change it.  The closest C# could come is with static methods and faking instance syntax, as it does with extension methods.  But static methods cannot participate in virtual dispatch meaning that they cannot be overridden by a derived class.

**Comment by VanKrock** [comment]
@HaloFour  It's syntax shugar. In IL method seems like
```csharp
        public TObj Set<TObj>(TObj @this, T value) where TObj : Obj
        {
            this.value = value;
            return @this;
        }
```

It's can be convert on compile?

**Comment by HaloFour** [comment]
@VanKrock

> It's can be convert on compile?

Not to a proper instance method, not without CLR changes.

There are other proposals to enable returning the same type as the instance type which I'll link to when I'm in front of a computer.

**Comment by Joe4evr** [comment]
#169 and #252, mostly.

If what you wanted is a `this` return type, you should've said so from the beginning. This is a classic case of [XY Problem](http://xyproblem.info/).

**Comment by HaloFour** [comment]
@Joe4evr 

Thanks.

> If what you wanted is a `this` return type, you should've said so from the beginning.

Agreed, much better to communicate a problem and then propose possible solutions.  Then any discussion can at least try to focus on solving the original problem.

**Comment by VanKrock** [comment]
@Joe4evr in issues #169 and #252 it's XY Problem, but in this solution it's not YX Problem
I can't to do fluent interfaces for generics because a can't to do ` thistype Set(T value)`. It's impossible. It's my problem.

But if I use  ` thistype Set(T value)` I must to resolve what is thistype. It's problem.
 ` TObj Set<TObj>(TObj obj, T value)` - valid  signature.
 ` TObj Set<TObj>(this TObj obj, T value)` - "this" means that obj is obj who called Set only. Similar static extension method. This signature valid for static class

**Comment by bondsbw** [comment]
@VanKrock I don't think you understand what the XY Problem is.  Your last comment would be a good example of it.

1. You want `thistype Set(T value)`
2. You can't do `thistype`, but might be able to get close enough with `TObj Set<TObj>(this TObj obj, T value)`
3. You can't do that either, so you ask for it to be a new feature
4. People question why you want it
5. It finally becomes clear that what you really wanted all along is `thistype`

**Comment by tannergooding** [comment]
> You can not return instance of Deprecated class from Base class.

@VanKrock, do you mean `Derived` class? A `Deprecated` class is one that you should no longer use

My understanding of your problem is that for:
```C#
public class MyClass
{
    public MyClass(string name)
    {
        Name = name;
    }

    public string Name { get; }

    public MyClass WithName(string name)
    {
        return new MyClass(name);
    }
}

public class MyDerivedClass : MyClass
{
    public MyDerivedClass(string name) : base(name)
    {
    }
}
```

Someone who uses `MyDerivedClass` can no longer call `WithName` fluently as it returns `MyClass` instead of `MyDerivedClass`.

Is that correct?

**Comment by VanKrock** [comment]
@bondsbw I want fluent interfaces but it's impossible in C# now

**Comment by VanKrock** [comment]
@tannergooding No. I want
```csharp
public class Customer
{
    private string name;
    public string Name => name;

    public MyClass WithName(string name)
    {
        this.name = name;
        return this;
    }
}

public class CustomerWithAge: Customer
{

    private int age;
    public int Age => age;
    public CustomerWithAge WithAge(int age)
    {
        this.age= age;
        return this;
    }
}

public class Test
{
    public void Run()
    {
        var customer = new CustomerWithAge()
            .WithName("CustomerName")
            .WithAge(20);
    }
}
```

**Comment by bondsbw** [comment]
We understood what you are asking for.
